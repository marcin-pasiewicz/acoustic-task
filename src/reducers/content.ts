import { ActionsTypes } from '../actions';

const initialState = {
    isFetchingContent: false,
    data: {},
    error: null,
};


export const content = (state = initialState, action: any) => {
    switch (action.type) {
        case ActionsTypes.FETCH_CONTENT:
            return {
                ...state,
                isFetchingContent: true,
            }
        case ActionsTypes.FETCH_CONTENT_FULFILLED:
            return {
                ...state,
                isFetchingContent: false,
                data: action.payload,
            }
        case ActionsTypes.FETCH_CONTENT_REJECTED:
            return {
                ...state,
                isFetchingContent: false,
                error: action.error,
            }
        default:
            return state
    }
}
