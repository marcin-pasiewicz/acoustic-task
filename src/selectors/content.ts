import { State } from "../models/State";

export function getContent(state: State) {
  return state;
}

export function getContentData(state: State) {
  return getContent(state).data;
}

export function isFetchingContent(state: State) {
  return getContent(state).isFetchingContent;
}

export function getContentError(state: State) {
  return getContent(state).error;
}
