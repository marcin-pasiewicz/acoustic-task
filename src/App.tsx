import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchContent } from './actions';
import { NoData } from './components/NoData/NoData'
import { getContentData, getContentError, isFetchingContent } from './selectors/content';
import Elements from './components/Elements/Elements';
import Loader from './components/Loader/Loader';
import ErrorBoundary from './components/ErrorBoundary/ErrorBoundary';
import { isEmpty } from './helpers/objectHelpers';

export const App: React.FC = () => {
  const dispatch = useDispatch();
  const content = useSelector(getContentData);
  const error = useSelector(getContentError);
  const isFetching = useSelector(isFetchingContent);
  const { elements = {} } = content;

  useEffect(() => {
    dispatch(fetchContent('fa9519d5-0363-4b8d-8e1f-627d802c08a8'))
  }, [dispatch])

  if (error) {
    return <NoData text="Ooops. That is not good" />
  }

  if (isFetching) {
    return <Loader />
  }

  return (
    <ErrorBoundary>
      {isEmpty(elements)
        ? <NoData text="Ooops. No data to show" />
        : <Elements elements={elements} />
      }
    </ErrorBoundary>
  );
}
