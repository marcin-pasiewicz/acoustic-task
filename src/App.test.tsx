import React from 'react';
import { render } from '@testing-library/react';
import { App } from './App';
import * as contetnSelectors from './selectors/content';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(fn => fn()),
  useDispatch: () => jest.fn(),
}));

const testData = {
  elements: { "heading": { "elementType": "text", "value": "Staffordshire Terrier Pups" }}
}

const setup = ({ isFetching = false, data = testData, error = null } = {}) => {
  jest.spyOn(contetnSelectors, 'getContentData').mockReturnValue(data);
  jest.spyOn(contetnSelectors, 'getContentError').mockReturnValue(error);
  jest.spyOn(contetnSelectors, 'isFetchingContent').mockReturnValue(isFetching);
};

describe('<App/>', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('renders correctly with data', () => {
    setup()
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders correctly with error', () => {
    setup({error: 'error'})
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders correctly with loader', () => {
    setup({isFetching: true})
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });
});
