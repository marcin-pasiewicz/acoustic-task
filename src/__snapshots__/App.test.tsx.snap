// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App/> renders correctly with data 1`] = `
<DocumentFragment>
  <div
    class="sc-AxiKw flXjnQ"
  >
    <div>
      <h1>
        Staffordshire Terrier Pups
      </h1>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<App/> renders correctly with error 1`] = `
<DocumentFragment>
  <div
    class="sc-AxjAm hvoVVA"
  >
    <svg
      height="82"
      viewBox="0 0 64 41"
      width="128"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        fill="none"
        fill-rule="evenodd"
        transform="translate(0 1)"
      >
        <ellipse
          cx="32"
          cy="33"
          fill="#F5F5F5"
          rx="32"
          ry="7"
        />
        <g
          fill-rule="nonzero"
          stroke="#D9D9D9"
        >
          <path
            d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
          />
          <path
            d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
            fill="#FAFAFA"
          />
        </g>
      </g>
    </svg>
    <h4>
      Ooops. That is not good
    </h4>
  </div>
</DocumentFragment>
`;

exports[`<App/> renders correctly with loader 1`] = `
<DocumentFragment>
  <svg
    display="block"
    height="200"
    preserveAspectRatio="xMidYMid"
    style="margin: auto; background-color: rgb(255, 255, 255);"
    viewBox="0 0 100 100"
    width="200"
    xmlns="http://www.w3.org/2000/svg"
  >
    <rect
      fill="#7a7a7a"
      height="12"
      rx="3"
      ry="3.6"
      width="6"
      x="47"
      y="24"
    >
      <animate
        attributeName="opacity"
        begin="-0.9166666666666666s"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        values="1;0"
      />
    </rect>
    <rect
      fill="#7a7a7a"
      height="12"
      rx="3"
      ry="3.6"
      transform="rotate(30 50 50)"
      width="6"
      x="47"
      y="24"
    >
      <animate
        attributeName="opacity"
        begin="-0.8333333333333334s"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        values="1;0"
      />
    </rect>
    <rect
      fill="#7a7a7a"
      height="12"
      rx="3"
      ry="3.6"
      transform="rotate(60 50 50)"
      width="6"
      x="47"
      y="24"
    >
      <animate
        attributeName="opacity"
        begin="-0.75s"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        values="1;0"
      />
    </rect>
    <rect
      fill="#7a7a7a"
      height="12"
      rx="3"
      ry="3.6"
      transform="rotate(90 50 50)"
      width="6"
      x="47"
      y="24"
    >
      <animate
        attributeName="opacity"
        begin="-0.6666666666666666s"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        values="1;0"
      />
    </rect>
    <rect
      fill="#7a7a7a"
      height="12"
      rx="3"
      ry="3.6"
      transform="rotate(120 50 50)"
      width="6"
      x="47"
      y="24"
    >
      <animate
        attributeName="opacity"
        begin="-0.5833333333333334s"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        values="1;0"
      />
    </rect>
    <rect
      fill="#7a7a7a"
      height="12"
      rx="3"
      ry="3.6"
      transform="rotate(150 50 50)"
      width="6"
      x="47"
      y="24"
    >
      <animate
        attributeName="opacity"
        begin="-0.5s"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        values="1;0"
      />
    </rect>
    <rect
      fill="#7a7a7a"
      height="12"
      rx="3"
      ry="3.6"
      transform="rotate(180 50 50)"
      width="6"
      x="47"
      y="24"
    >
      <animate
        attributeName="opacity"
        begin="-0.4166666666666667s"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        values="1;0"
      />
    </rect>
    <rect
      fill="#7a7a7a"
      height="12"
      rx="3"
      ry="3.6"
      transform="rotate(210 50 50)"
      width="6"
      x="47"
      y="24"
    >
      <animate
        attributeName="opacity"
        begin="-0.3333333333333333s"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        values="1;0"
      />
    </rect>
    <rect
      fill="#7a7a7a"
      height="12"
      rx="3"
      ry="3.6"
      transform="rotate(240 50 50)"
      width="6"
      x="47"
      y="24"
    >
      <animate
        attributeName="opacity"
        begin="-0.25s"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        values="1;0"
      />
    </rect>
    <rect
      fill="#7a7a7a"
      height="12"
      rx="3"
      ry="3.6"
      transform="rotate(270 50 50)"
      width="6"
      x="47"
      y="24"
    >
      <animate
        attributeName="opacity"
        begin="-0.16666666666666666s"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        values="1;0"
      />
    </rect>
    <rect
      fill="#7a7a7a"
      height="12"
      rx="3"
      ry="3.6"
      transform="rotate(300 50 50)"
      width="6"
      x="47"
      y="24"
    >
      <animate
        attributeName="opacity"
        begin="-0.08333333333333333s"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        values="1;0"
      />
    </rect>
    <rect
      fill="#7a7a7a"
      height="12"
      rx="3"
      ry="3.6"
      transform="rotate(330 50 50)"
      width="6"
      x="47"
      y="24"
    >
      <animate
        attributeName="opacity"
        begin="0s"
        dur="1s"
        keyTimes="0;1"
        repeatCount="indefinite"
        values="1;0"
      />
    </rect>
  </svg>
</DocumentFragment>
`;
